enhancements/storage/csi-efs-operator.md

```diff
@@ -0,0 +1,335 @@
+
+* It's written using `controller-runtime`.
+* Once the operator is installed by OLM, the operator automatically installs the CSI driver, without any CRD/CR.
+  * As consequence, user cannot un-install the CSI driver easily.
+* By default, it is installed in `openshift-operators` namespace and the CSI driver runs there too.
+* It does not create EFS volumes in AWS! It's up to the cluster admin to create EFS volume in AWS, figure in which VPC
+  it should be available and set correct security group. See [this KB article](https://access.redhat.com/articles/5025181)
+  for suggested procedure.
+* It offers `SharedVolume` CRD in `aws-efs.managed.openshift.io/v1alpha1` API group to create PV + PVC for an EFS
+  volume. The EFS volume still needs to be created manually, see the previous bullet!
+  * All that `SharedVolume` really does is that it creates PV with proper settings + a PVC. This allows unprivileged
+    users to create PVs. Cluster admins need to be careful to give permissions to use this CRD only to trusted users.
+  * Using `SharedVolume` is purely optional! Cluster admin may create PVs for EFS volumes manually, it's not harder
+    than using `SharedVolume`.
+
+### AWS EFS CSI driver
+
+AWS EFS CSI driver is different than the other CSI drivers we ship. Listing the main differences here for completeness.
+
+* Right now, it does not implement dynamic provisioning. As described above, cluster admin must create the volume in
+  AWS and figure out the networking + security manually.
+  * [Long upstream thread about dynamic provisioning](https://github.com/kubernetes-sigs/aws-efs-csi-driver/issues/6)
+  * Upstream [will offer some sort of dynamic provisioning in the next version](https://github.com/kubernetes-sigs/aws-efs-csi-driver/pull/274)
+    by creating subdirectories (Access Points) in existing EFS volumes as PVs. AWS has a limit of 120 Access Points per
+    EFS volume, i.e. only 120 PVs can be created for a single StorageClass.
+    [Upstream issue](https://github.com/kubernetes-sigs/aws-efs-csi-driver/issues/310).
+  * Proper dynamic provisioning can come in the future, dealing with the networking and security groups is the hardest
+    part there.
+* AWS EFS volumes do not have any size enforcement. They can hold as much data as apps put there and someone will have to
+  pay for the storage. OCP offer standard metrics for amount of data on PVCs, it's up to the cluster admin to create
+  alerts if the consumption gets too high.
+* It uses [efs-utils](https://github.com/aws/efs-utils) to provide encryption in transport, which is not included in
+  RHEL. We will need to package it, either as an image or RPM package.
+
+## Proposal
+High level only, see design details below.
+
+1. Rewrite current aws-efs-operator to use [CSI driver installation functions of library-go](https://github.com/openshift/library-go/tree/master/pkg/operator/csi/csicontrollerset).
```

Comment:
Since you are already using controller-runtime, have you considered making this an [operator-sdk](https://github.com/operator-framework/operator-sdk/) project? We have [integration with OLM](https://master.sdk.operatorframework.io/docs/olm-integration/) like `operator-sdk run bundle` which will help you test out your bundle with OLM, among other things. 

There's also a supported operator-sdk offering beginning with OpenShift 4.7:
https://mirror.openshift.com/pub/openshift-v4/clients/operator-sdk/latest/

Reply:
I tried to cover *why* in Design Details below. This is almost complete CSI driver operator using library-go code:

https://github.com/openshift/aws-ebs-csi-driver-operator/blob/383a7638b37a4b9a9831c7747e8c499eedcf030f/pkg/operator/starter.go

Basically configure a library and run it. It's so comfortable. We would need to re-implement it in operator-sdk and then we would have two code bases that we need to maintain and sync.

Reply:
I'll just point out that aws-efs-operator *is* written against operator-sdk (but 0.16.0, so no `run bundle`). Jan has been talking about porting to library-go for $reasons. To editorialize shamelessly, I feel like it's because it's what he's used to: osdk/controller-runtime is likewise "so comfortable" for me :)